[usage-collection] CLI for host usage collection

AparnaKarve opened the pull request:
https://issues.redhat.com/browse/AA-1576

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Extended the functionality of the `host_metrics` CLI as described in the ADR

It now supports -

```awx-manage host_metric --csv host_metric```
```awx-manage host_metric --csv host_metric_summary_monthly```

 The above options would output the contents of the CSV in the standard output


```awx-manage host_metric --tarball```

```awx-manage host_metric  --tarball --rows_per_file ``` 

The above options would produce a tarball

For e.g - 
```awx-manage host_metric  --tarball --rows_per_file 100000 ``` 


-----------------------

Sample output with the `--csv` option :
```
# awx-manage host_metric --csv host_metric
id,hostname,first_automation,last_automation,last_deleted,automated_counter,deleted_counter,deleted,used_in_inventories
1,localhost,2023-01-18 17:53:15.824452+00:00,2023-01-18 17:53:15.823961+00:00,,0,0,False,
3,awx_1,2023-01-18 18:28:40.421047+00:00,2023-01-18 19:12:34.620062+00:00,,0,0,False,
2,awx_1_copy,2023-01-18 18:32:57.004374+00:00,2023-01-18 18:32:57.003868+00:00,,0,0,False,
5,host1,2023-01-18 22:49:32.891346+00:00,2023-02-09 19:51:37.351248+00:00,,4,0,False,
4,host2,2023-01-18 22:49:32.891430+00:00,2023-02-09 19:51:37.351248+00:00,,4,0,False,
6,host2a,2023-02-02 20:14:48.108894+00:00,2023-02-09 19:51:37.351248+00:00,,3,0,False,
14,host4,2023-02-04 01:47:36.704301+00:00,2023-02-15 16:55:37.932600+00:00,,4,0,False,
15,host3,2023-02-04 01:47:36.704358+00:00,2023-02-15 16:55:37.932600+00:00,,4,0,False,
20,host9,2023-02-06 23:00:45.343491+00:00,2023-02-09 19:51:37.351248+00:00,,4,0,False,
21,host8,2023-02-06 23:00:45.343556+00:00,2023-02-09 19:51:37.351248+00:00,,4,0,False,
36,host5,2023-02-09 07:10:44.545754+00:00,2023-02-15 16:55:37.932600+00:00,,4,0,False,
37,host7,2023-02-09 07:10:44.545814+00:00,2023-02-15 16:55:37.932600+00:00,2023-02-15 16:54:57.964919+00:00,4,1,False,
38,host6,2023-02-09 07:10:44.545857+00:00,2023-02-15 16:55:37.932600+00:00,,4,0,False,
57,abc,2023-02-09 17:49:05.996232+00:00,2023-02-09 19:11:46.847399+00:00,2023-02-15 16:49:48.252720+00:00,1,1,True,
88,host10,2023-02-14 02:20:56.119105+00:00,2023-02-15 16:20:16.958107+00:00,,2,0,False,
```




<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New or Enhanced Feature

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - CLI


##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 21.12.1.dev1+g6ea81c4ec9
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
All anonymization code that previously existed has been removed

~For the SHA256 anonymization, we could not completely leverage the existing `awx.main.utils.encryption` package because it looks like the `decrypt_value` function does not respond well if a custom salt is used during encryption.
For that reason, we had to write our own `encrypt_name` and `decrypt_name` (the way we have it in Analytics)
Note: While both functions, `encrypt_name` and `decrypt_name` are defined, for this CLI utility,  only `encrypt_name` has been used.~

AparnaKarve committed:
CLI for host usage collection

Ladas commented on awx/main/management/commands/host_metric.py:
```diff
@@ -1,26 +1,196 @@
+            f.write(config_json)
+        return config_file, arcname_file
+
+    def encrypt_name(self, key, value):
```
btw. the ADR had it needs Anonymization not Encryption (these are 2 different approaches, with anonymization being irreversible)

But it might be best to remove anonymization for now, we will need a proper design for how to deal with a drift if key is changed. (otherwise it will create chaos in the reporting)

AparnaKarve replied:
So, remove the support for `--anonymized` for now?

Ladas replied:
@AparnaKarve yeah that might be the best, we can try to solve it for 2.5 in general (so not just for CLI), with a plan of how to deal with the drift.

TheRealHaoLiu commented:
summary for the code review session:
- PR and functionality looks good 
- looks like pandas is not being used by anything critical and the csv library would be enough, we would like to avoid pulling in pandas if possible
- removing the anonymizing function due to comment from @Ladas

AparnaKarve committed:
Remove `pandas` use `csv`. Also, remove anonymization

AparnaKarve commented:
Addressed above feedback.

Previously, `pandas` was used for CSV creation and for applying a column-wide anonymization function.
With the plan to remove anonymization for 2.4, it made sense to use the `csv` library (already part of awx) for CSV creation.

slemrmartin commented on awx/main/management/commands/host_metric.py:
```diff
@@ -33,17 +174,39 @@ def handle(self, *args, **options):
+        host_metric_row_count = HostMetric.objects.filter(**filter_kwargs).count()
+        host_metric_summary_monthly_row_count = HostMetricSummaryMonthly.objects.filter(**filter_kwargs_host_metrics_summary).count()
+
+        if (host_metric_row_count > PREFERRED_ROW_COUNT or host_metric_summary_monthly_row_count > PREFERRED_ROW_COUNT) and (
```
I suggest to set the 500k as a default value (ADR is describing it too). The `--rows_per_file` can be optional customization.

AparnaKarve replied:
The default value is already set to 500k.
We can set `--rows_per_file` to be an optional customization -- that's fine.

Note: This will require a change in the ADR

Ladas replied:
I think ADR lists that as optional (stating 500k should be default)

slemrmartin commented on awx/main/management/commands/host_metric.py:
```diff
@@ -33,17 +174,39 @@ def handle(self, *args, **options):
+            self.output_csv(options, filter_kwargs)
+        elif options['tarball']:
+            self.output_tarball(options, filter_kwargs, host_metric_row_count, host_metric_summary_monthly_row_count)
+        elif options['rows_per_file']:
```
as this param could be just a customization for csv/tarball options, it shouldn't be in the if/else flow

slemrmartin commented on awx/main/management/commands/host_metric.py:
```diff
@@ -1,26 +1,167 @@
+    def paginated_db_retrieval(self, type, filter_kwargs, offset=0, limit=PREFERRED_ROW_COUNT):
+        list_of_queryset = []
+        if type == 'host_metric':
+            result = HostMetric.objects.filter(**filter_kwargs)
```
there could be also fine-grained db pagination in case of big rows_per_file settings (i.e. 10000)

This can produce `SELECT * from main_hostmetric WHERE ... LIMIT 500000`. That's quite a big number for db.

AparnaKarve replied:
So `--rows_per_file`  is the parameter for regulating how many rows to fetch at any given time.



- `For awx-manage host_metric --csv --rows_per_file 10000`, 
we will keep fetching 10000 rows until we have fetched all rows and then display the csv in standard output

- `For awx-manage host_metric --tarball --rows_per_file 10000`, 
we will fetch 10000 rows and create the tarball with several individual CSV files where each CSV would have up to 10000 rows

- `For awx-manage host_metric --tarball `, 
we will fetch 500k rows in one go since `--rows_per_file` was not specified and create the tarball

slemrmartin replied:
yes, it's a parameter how big should be the CSV file. 
But my concern is that if we default 500k, then *default* query to the database is `SELECT .. LIMIT 500k`, that sounds dangerous. So I'm suggesting to make independent db pagination and csv pagination

Ladas replied:
Martin is correct, unless you'll do `COPY` directly to csv (that should be batched operation), this might blow up in memory. E.g. we do this in Analytics Exporter, COPY to csv without any explicit batching (postgres takes care of that), then batch reading of the csv to convert it to different format. (so we keep the memory low) That should be the most effective way to do it.


But paginated read from DB should work ok-ish for now. (something like hardcoded page of 10k rows we load from DB at once and append to the csv)

AparnaKarve replied:
> So I'm suggesting to make independent db pagination and csv pagination

I had thought about this, but was trying to avoid this situation because two pagination schemes are just going to add some complexity.

I would have to evaluate this to make a decision as we do not want to overcomplicate the code.

------------------

The easiest approach here  would be to set default value of `--rows_per_file` to 10000

AparnaKarve replied:
Added functionality for fetching rows in batches of 10k and creating CSVs with whatever is specified in `--rows_per_file`
Since 500k is a really large number (even for the postgres function), its best to use 10k-chunking approach.

(Also, the postgres function might force us to use raw SQL - would prefer to use django orm when possible)

slemrmartin commented on awx/main/management/commands/host_metric.py:
```diff
@@ -33,17 +174,39 @@ def handle(self, *args, **options):
+        if until is not None:
+            filter_kwargs_host_metrics_summary['date__lte'] = until
+
+        host_metric_row_count = HostMetric.objects.filter(**filter_kwargs).count()
```
these 2 db consuming queries can be omitted if there will be loop with increasing db offset and checked emptiness of retrieved queries.

AparnaKarve replied:
This should be doable.

slemrmartin reviewed (changes requested):
Overall it looks good.
The pagination parts and pagination param should be adjusted and it'll be ready

AparnaKarve commented:
@Ladas @slemrmartin If the [above](https://github.com/ansible/awx/pull/13585#discussion_r1123414850) resolution looks OK, please update the ADR and I will make the requested changes.
Thanks.

AparnaKarve committed:
make `rows_per_file` optional parameter

Removed 2 sql statements that gave the info on row count
which warranted many other changes

AparnaKarve commented:
Addressed the above feedback related to -
- making `--rows_per_file` optional
- Not calling the `.count` queries for the 2 tables and instead relying on the empty queryset list to determine if we have fetched all the rows from the db

AparnaKarve committed:
Added internal batch processing for up to 10k rows

For --rows_per_file if > 10k, rows would be fetched in batches of 10k

slemrmartin reviewed (approved):
Looks good now, thanks :+1:
